File: server/src/chains/basic-chat.js
/**
 * Chain 链式调用
 * 使用 LCEL（LangChain Expression Language）管道语法
 * 将 Prompt → Model → OutputParser 串联
 * LCEL 数据流：输入对象 → Prompt 格式化 → messages 数组 → Model 调用 → AIMessage → Parser 提取 → 字符串
 *
 * 数据流说明：
 * 1. invoke({ chat_history, user_input, current_time }) 传入 Prompt 模板
 * 2. Prompt 模板格式化为 messages 数组，传入 Model
 * 3. Model 调用 DeepSeek API，返回 AIMessage 对象
 * 4. StringOutputParser 从 AIMessage 中提取纯文本字符串
 */

import { StringOutputParser } from "@langchain/core/output_parsers";
import { createModel } from "../models/deepseek.js";
import {
  customerServicePrompt,
  generalChatPrompt,
} from "../prompts/customer-service.js";

/**
 * @description 极速购客服 Chain（非流式）
 */
const model = createModel({ temperature: 0.5 });
const parser = new StringOutputParser();

// Prompt → Model → Parser
export const customerServiceChain = customerServicePrompt
  .pipe(model)
  .pipe(parser);

/**
 * @description 极速购客服 Chain（流式）
 */
const streamingModel = createModel({ temperature: 0.5, streaming: true });
export const customerServiceStreamChain = customerServicePrompt
  .pipe(streamingModel)
  .pipe(parser);

/**
 * @description 通用对话 Chain（演示用）
 */
export const generalChatChain = generalChatPrompt.pipe(model).pipe(parser);

/**
 * @description 工具函数：格式化历史消息
 * 将前端传来的 { role, content } 数组转换为 LangChain 消息格式
 * LangChain 支持 human / assistant / system 三种 role
 */
export const formatHistory = (history = []) => {
  return history
    .map((msg) => {
      if (msg.role === "user") return ["human", msg.content];
      if (msg.role === "assistant") return ["assistant", msg.content];
      return null;
    })
    .filter(Boolean);
};

File: backend_py/chains/basic_chat.py
"""
基础对话 Chain
使用 LCEL 管道语法：Prompt | Model | Parser

对照 JS: customerServicePrompt.pipe(model).pipe(parser)
Python:  customer_service_prompt | model | parser
"""
from langchain_core.output_parsers import StrOutputParser
from langchain_core.messages import HumanMessage, AIMessage
from models.deepseek import create_model
from prompts.customer_service import customer_service_prompt, general_chat_prompt

# 注意：Python 中是 StrOutputParser，不是 StringOutputParser
parser = StrOutputParser()

# 非流式 Chain
model = create_model(temperature=0.5)
customer_service_chain = customer_service_prompt | model | parser

# 流式 Chain
streaming_model = create_model(temperature=0.5, streaming=True)
customer_service_stream_chain = customer_service_prompt | streaming_model | parser

# 通用对话 Chain
general_chat_chain = general_chat_prompt | model | parser


def format_history(history=None):
    """
    工具函数：格式化历史消息

    对照 JS:
      export const formatHistory = (history = []) => {
        return history.map(msg => {
          if (msg.role === "user") return ["human", msg.content]
          if (msg.role === "assistant") return ["assistant", msg.content]
          return null
        }).filter(Boolean)
      }

    Python 差异：返回 Message 对象列表，而不是元组
    """
    if history is None:
        history = []

    messages = []
    for msg in history:
        if msg.get("role") == "user":
            messages.append(HumanMessage(content=msg["content"]))
        elif msg.get("role") == "assistant":
            messages.append(AIMessage(content=msg["content"]))
    return messages